### ISSUE TITLE
Problems with Notebook debugger (using Jupyter Protocol)

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode-jupyter#10106]
**Note: These issues have been fixed for IW debugger, we need to fix them for Notebook**

* Problem 1:
	* Cell 1
	  ```python
	  < empty line>
	  < empty line>	  
	  < empty line>
	  < empty line>
	  print(1)
	  print(2)
	  ```
	When debugging the above code, the debugger will stop at line `print(1)`, however when you hit `step over` in RBL or `continue` in the cell debugger it will break there once again.

* Problem 2:
	* Cell 1
	  ```python
	  < empty line>
	  < empty line>	  
	  < empty line>
	  def foo():
	      print('inside foo')
	  ```
	  * Cell 2
	  ```python
	  
	  
	  foo()
	  ```
	  When debugging the above code, the debugger will stop at line `foo()`, however when you step into the code, it will stop way above the `def foo`, in a white space (basically on line 2 of the file, in the white space).
	  

* Problem 3:
	The above issues were uncovered while running some of the IW debugger tests.
	I think we need a few more tests for the notebook debugger to cover the same scenarios covered in IW debugger.

These issues have been fixed in the new IW when using debugger protocol.
Basically injecting `breakpiont()` along with a few other changes.

**Comment by DonJayamanne** [comment]
After further analysis, i think we have to do the following:
* Use `breakpoint()` to stop where we want (instead of adding breakpoints)
* Strip the leading white space from the cells and send that to the kernel (debugger)
	* We'll need to take this into account when translating source maps & tracebacks

We might want to make this change for Notebook debugger as well.

**Comment by roblourens** [comment]
I can't repro issue 1 in a notebook. I see issue 2 though, didn't know that Jupyter does this

**Comment by DonJayamanne** [comment]
@roblourens the empty lines were stripped during rendering... lol, Jupyter does the same, 
Try with the leading empty lines.

**Comment by ttSpace** [comment]
Build of Visual Studio Code - Insiders: 1.68.0-insider(user setup) 
Build of python-Pre-Release: v2022.7.11541009 
Build of jupyter-Pre-Release: v2022.6.1001562429 
Build of pylance: v2022.6.0 
Build of jupyter Notebook Renderers: v1.0.8 
Build of jupyter Keymap: v1.0.0 

Pressing F10 will stop at the line ```def bar():```, then pressing F10 again will stop at the line ``def bar():```.

![image](https://user-images.githubusercontent.com/82852072/172128586-14e2bfe8-5971-4bb0-acab-4802b6eba227.png)

**Comment by DonJayamanne** [comment]
@ttSpace
please could you file a separate issue for that.

**Comment by DonJayamanne** [comment]
Upstream issue https://github.com/microsoft/vscode-jupyter/issues/10586

**Comment by DonJayamanne** [comment]
Issues with empty lines will not be resolved for now (as it was reported by me and a papercut), 
Creating a new issue to track breakpoint issue caused by upstream python debgger https://github.com/microsoft/vscode-jupyter/issues/10600

**Comment by roblourens** [comment]
Was problem 2 supposed to be fixed by the debugpy update?

<img width="619" alt="image" src="https://user-images.githubusercontent.com/323878/185268440-1af78b65-7b79-4d63-a8a8-2b6176b538b9.png">
